promql/query_logger.go

```diff
@@ -0,0 +1,208 @@
+	err := json.Indent(&out, []byte(queries), "", "\t")
+
+	if err != nil {
+		level.Warn(logger).Log("msg", "Error parsing json", "err", err)
+		return false, "[]"
+	}
+
+	return true, out.String()
+}
+
+func logUnfinishedQueries(filename string, filesize int, logger log.Logger) {
+	if _, err := os.Stat(filename); err == nil {
+		fd, err := os.Open(filename)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to open query log file", "err", err)
+			return
+		}
+
+		brokenJson := make([]byte, filesize)
+		_, err = fd.Read(brokenJson)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to read query log file", "err", err)
+			return
+		}
+
+		queriesExist, queries := parseBrokenJson(brokenJson, logger)
+		if !queriesExist {
+			return
+		}
+		level.Info(logger).Log("msg", "These queries didn't finish in prometheus' last run:")
+		fmt.Println(queries)
+	}
+}
+
+func getMMapedFile(filename string, filesize int, logger log.Logger) (error, []byte) {
+
+	fd, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error opening query log file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	if err := fd.Truncate(int64(filesize)); err != nil {
+		level.Error(logger).Log("msg", "Error creating file of specified size", "file", filename, "Attemped size", filesize, "err", err)
+		return err, []byte{}
+	}
+
+	fileAsBytes, err := syscall.Mmap(int(fd.Fd()), 0, filesize, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error mmapping file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	return err, fileAsBytes
+}
+
+func NewActiveQueryTracker(maxQueries int, logger log.Logger) (activeQueryTracker ActiveQueryTracker) {
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(logger).Log("msg", "Panic while creating ActiveQueryTracker", "err", r)
+			activeQueryTracker = ActiveQueryTracker{}
+		}
+	}()
+
+	filename, filesize := "queries.active", 1+maxQueries*entrySize
+	logUnfinishedQueries(filename, filesize, logger)
+
+	err, fileAsBytes := getMMapedFile(filename, filesize, logger)
+	if err != nil {
+		return ActiveQueryTracker{}
+	}
+
+	copy(fileAsBytes, "[")
+	activeQueryTracker = ActiveQueryTracker{
+		mmapedFile:   fileAsBytes,
+		getNextIndex: make(chan int, maxQueries),
+		lastIndex:    make(chan int, (filesize-1)/entrySize),
+		logger:       logger,
+	}
+
+	activeQueryTracker.generateIndices(filesize)
+
+	return activeQueryTracker
+}
+
+type Entry struct {
+	Query     string `json:"query"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func trimStringByBytes(str string, size int) string {
+	bytesStr := []byte(str)
+	return string(bytesStr[:min(len(bytesStr), size)])
+}
+
+func newJsonEntry(query string, timestamp int64, logger log.Logger) []byte {
+	// 35 bytes is the size of json entry for empty string.
```

Comment:
Wouldn't that vary based on the timestamp?

Reply:
I didn't know about that. I assumed all unix timestamps would be 10 characters long. I'll add a function to dynamically compute max byte length possible for the string.